### PR TITLE
Tangram 2.2

### LINKED ISSUE #556: Upgrade Tangram to version 0.2.2

### PR DESCRIPTION
**Pull request by sarahlensing**
Closes #556 #560 

Begins migration from MapController to MapzenMap

### CODE REVIEW

**Comment by ecgreb** [comment]
I am getting a crash on first launch after installing this version. It appears `OverlayManager` is attempting to access location before I have finished agreeing to the runtime permission dialog.

```
05-18 14:33:39.651 15162-15162/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                   Process: com.mapzen.erasermap, PID: 15162
                                                   java.lang.SecurityException: "passive" location provider requires ACCESS_FINE_LOCATION permission.
                                                       at android.os.Parcel.readException(Parcel.java:1620)
                                                       at android.os.Parcel.readException(Parcel.java:1573)
                                                       at android.location.ILocationManager$Stub$Proxy.getLastLocation(ILocationManager.java:717)
                                                       at android.location.LocationManager.getLastKnownLocation(LocationManager.java:1214)
                                                       at com.mapzen.android.lost.internal.FusionEngine.getLastLocation(FusionEngine.java:50)
                                                       at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.getLastLocation(FusedLocationProviderApiImpl.java:32)
                                                       at com.mapzen.android.OverlayManager.showLastKnownLocation(OverlayManager.java:412)
                                                       at com.mapzen.android.OverlayManager.handleMyLocationEnabledChanged(OverlayManager.java:380)
                                                       at com.mapzen.android.OverlayManager.setMyLocationEnabled(OverlayManager.java:137)
                                                       at com.mapzen.android.MapzenMap.setMyLocationEnabled(MapzenMap.java:229)
                                                       at com.mapzen.erasermap.controller.MainActivity.configureMapzenMap(MainActivity.kt:311)
                                                       at com.mapzen.erasermap.controller.MainActivity.access$configureMapzenMap(MainActivity.kt:86)
                                                       at com.mapzen.erasermap.controller.MainActivity$initMapzenMap$1.onMapReady(MainActivity.kt:255)
                                                       at com.mapzen.android.MapInitializer$1.onMapReady(MapInitializer.java:75)
                                                       at com.mapzen.tangram.MapView$1.onPostExecute(MapView.java:77)
                                                       at com.mapzen.tangram.MapView$1.onPostExecute(MapView.java:65)
                                                       at android.os.AsyncTask.finish(AsyncTask.java:651)
                                                       at android.os.AsyncTask.access$500(AsyncTask.java:180)
                                                       at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:668)
                                                       at android.os.Handler.dispatchMessage(Handler.java:102)
                                                       at android.os.Looper.loop(Looper.java:148)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

File: app/build.gradle [hunk 1]
```diff
@@ -33,7 +33,7 @@ android {
 
   defaultConfig {
     applicationId "com.mapzen.erasermap"
-    minSdkVersion 15
+    minSdkVersion 17
```
Comment (ecgreb) [inline app/build.gradle]: Why the change for `minSdkVersion`?

Its probably fine to drop support for API 15 (2%) but there are still 7.2% of devices on API 16 currently.

**Comment by sarahlensing** [comment]
Nice catch. I've added checks to ensure we have permission before enabling location

File: app/build.gradle (hunk 1 above)
Comment (sarahlensing) [inline app/build.gradle]: Reverted, we don't need to change this value. In testing, I had layout param verbs that required this but then removed them.

File: app/src/main/kotlin/com/mapzen/erasermap/model/AndroidAppSettings.kt [hunk 2]
```diff
@@ -3,16 +3,17 @@ package com.mapzen.erasermap.model
 import android.content.Context
 import android.location.Location
 import android.preference.PreferenceManager
+import com.mapzen.android.MapzenMap
 import com.mapzen.erasermap.EraserMapApplication
 import com.mapzen.erasermap.R
 import com.mapzen.pelias.SavedSearch
-import com.mapzen.tangram.DebugFlags
-import com.mapzen.tangram.Tangram
+import com.mapzen.tangram.MapController
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/model/AndroidAppSettings.kt]: Unused import

Comment (sarahlensing) [inline app/src/main/kotlin/com/mapzen/erasermap/model/AndroidAppSettings.kt]: This is actually still used for the debug flags. Will be removed in an upcoming PR though

File: app/src/main/kotlin/com/mapzen/erasermap/controller/MainActivity.kt [hunk 3]
```diff
@@ -619,12 +602,93 @@ class MainActivity : AppCompatActivity(), MainViewController, RouteCallback,
         searchResultsView.onSearchResultsSelectedListener = this
     }
 
+    private fun baseAttributionParams(): RelativeLayout.LayoutParams {
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/controller/MainActivity.kt]: Wow this so much code to manipulate layout params. I wish there was a better way to do this in Android 😢 

One idea I had was to just adjust the layout params on `MapView` itself now that the find me button and attribution are part of it. But I am guessing that would have the side effect of changing the position of the map if we mess with the layout params on the view?

File: app/src/main/kotlin/com/mapzen/erasermap/presenter/MainPresenterImpl.kt [hunk 4]
```diff
@@ -315,28 +315,21 @@ public open class MainPresenterImpl(val mapzenLocation: MapzenLocation, val bus:
...
-        } else {
-            mainViewController?.showCurrentLocation(event.location)
         }
+        //TODO
+//        else {
+//            mainViewController?.showCurrentLocation(event.location)
+//        }
     }
 
     override fun onCreate() {
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/presenter/MainPresenterImpl.kt]: This method can be removed if its no longer needed.

File: app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocation.kt [hunk 5]
```diff
@@ -1,11 +1,12 @@
 package com.mapzen.erasermap.model
 
 import android.location.Location
+import com.mapzen.android.MapzenMap
 import com.mapzen.pelias.PeliasLocationProvider
 import com.mapzen.tangram.MapController
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocation.kt]: Unused import?

File: app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocationImpl.kt [hunk 6]
```diff
@@ -4,6 +4,7 @@ import android.content.Context
 import android.location.Location
 import android.util.Log
 import android.view.WindowManager
+import com.mapzen.android.MapzenMap
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocationImpl.kt]: `MapController` import can be removed here too.

File: app/src/main/kotlin/com/mapzen/erasermap/view/RouteModeView.kt [hunk 7]
```diff
@@ -250,17 +242,8 @@ class RouteModeView : LinearLayout, RouteViewController, ViewPager.OnPageChangeL
...
-            Tangram.addDataSource(routeIcon);
-        }
-
-        val properties = com.mapzen.tangram.Properties()
-        properties.set(MAP_DATA_PROP_TYPE, MAP_DATA_PROP_POINT);
-
-        routeIcon?.clearData()
-        routeIcon?.addPoint(properties, LngLat(location.longitude, location.latitude))
-        mapController?.requestRender()
+        mapzenMap?.clearRouteLocationMarker()
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/RouteModeView.kt]: Wow this method really cleaned up nice 👍

File: app/src/test/kotlin/com/mapzen/erasermap/model/TestAppSettings.kt [hunk 8]
```diff
@@ -2,8 +2,10 @@ package com.mapzen.erasermap.model
 
 import android.content.Context
 import android.location.Location
+import com.mapzen.android.MapzenMap
 import com.mapzen.erasermap.dummy.TestHelper
 import com.mapzen.pelias.SavedSearch
+import com.mapzen.tangram.MapController
```
Comment (ecgreb) [inline app/src/test/kotlin/com/mapzen/erasermap/model/TestAppSettings.kt]: Unused import? I sure wish Kotlin had a version of Checkstyle to validate this stuff for us :/

File: app/src/test/java/com/mapzen/erasermap/TestMapController.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import com.mapzen.tangram.LngLat;
+import com.mapzen.tangram.MapController;
+import com.mapzen.tangram.MapData;
+import com.mapzen.tangram.TestMapData;
+
+import android.opengl.GLSurfaceView;
+
+import java.util.HashSet;
+
+public class TestMapController extends MapController {
```
Comment (ecgreb) [inline app/src/test/java/com/mapzen/erasermap/TestMapController.java]: This and `TestMapData` are nicely designed test doubles ;) but I don't currently see them used anywhere. Is there a plan to add these to certain tests later? Or maybe do we not need them after all?

File: app/src/test/kotlin/com/mapzen/erasermap/model/TestMapzenLocation.kt [hunk 10]
```diff
@@ -1,14 +1,15 @@
 package com.mapzen.erasermap.model
 
 import android.location.Location
+import com.mapzen.android.MapzenMap
 import com.mapzen.pelias.BoundingBox
 import com.mapzen.tangram.MapController
```
Comment (ecgreb) [inline app/src/test/kotlin/com/mapzen/erasermap/model/TestMapzenLocation.kt]: Unused import?

File: app/src/main/kotlin/com/mapzen/erasermap/controller/MainActivity.kt (hunk 3 above)
Comment (sarahlensing) [inline app/src/main/kotlin/com/mapzen/erasermap/controller/MainActivity.kt]: I know, its pretty unfortunate :/

We could do this but it wouldn't save us much code and actually in the case where we have multiple search results (and we show the text indicator above our pager) we will have to still adjust the attribution text separately. We could/should add a method as part of the sdk to make this less painful for EM and for other developers though.

File: app/src/main/kotlin/com/mapzen/erasermap/presenter/MainPresenterImpl.kt (hunk 4 above)
Comment (sarahlensing) [inline app/src/main/kotlin/com/mapzen/erasermap/presenter/MainPresenterImpl.kt]: Rm'd

File: app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocation.kt (hunk 5 above)
Comment (sarahlensing) [inline app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocation.kt]: Rm'd

File: app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocationImpl.kt (hunk 6 above)
Comment (sarahlensing) [inline app/src/main/kotlin/com/mapzen/erasermap/model/MapzenLocationImpl.kt]: Rm'd

File: app/src/test/kotlin/com/mapzen/erasermap/model/TestAppSettings.kt (hunk 8 above)
Comment (sarahlensing) [inline app/src/test/kotlin/com/mapzen/erasermap/model/TestAppSettings.kt]: Rm'd

File: app/src/test/java/com/mapzen/erasermap/TestMapController.java (hunk 9 above)
Comment (sarahlensing) [inline app/src/test/java/com/mapzen/erasermap/TestMapController.java]: We don't need them- rm'd!

File: app/src/test/kotlin/com/mapzen/erasermap/model/TestMapzenLocation.kt (hunk 10 above)
Comment (sarahlensing) [inline app/src/test/kotlin/com/mapzen/erasermap/model/TestMapzenLocation.kt]: Rm'd
